engineapi/engineapi/middleware.py

```diff
@@ -199,3 +212,157 @@ async def dispatch(
+            timeout=BUGOUT_REQUEST_TIMEOUT_SECONDS,
+        )
+
+    except Exception as err:
+        logger.error(f"Error fetching bugout resources with CORS origins: {str(err)}")
+        return ALLOW_ORIGINS
+
+    # If there are no resources with CORS origins configuration, create new one
+    # with default list of origins from environment variable
+    if len(resources.resources) == 0:
+        try:
+            resource = bc.create_resource(
+                token=MOONSTREAM_ADMIN_ACCESS_TOKEN,
+                application_id=MOONSTREAM_APPLICATION_ID,
+                resource_data={
+                    "type": BUGOUT_RESOURCE_TYPE_APPLICATION_CONFIG,
+                    "setting": "cors",
+                    "user_id": str(MOONSTREAM_ADMIN_USER.id),
+                    "origins": ALLOW_ORIGINS,
+                },
+            )
+            resources.resources.append(resource)
+            logger.info(
+                "Created resource with default CORS origins setting by moonstream admin user"
+            )
+        except Exception as err:
+            logger.error(
+                f"Unable to write default CORS origins to resources: {str(err)}"
+            )
+            return ALLOW_ORIGINS
+
+    resource_origins_set: Set[str] = parse_origins_from_resources(resources.resources)
+    resource_origins_set = check_default_origins(resource_origins_set)
+
+    return list(resource_origins_set)
+
+
+def set_cors_origins_cache(allow_origins: Set[str]) -> None:
+    try:
+        allow_origins_str = ",".join(list(allow_origins))
+        rc_client.set("cors", allow_origins_str)
+    except Exception:
+        logger.warning("Unable to set CORS origins at Redis cache")
+    finally:
+        rc_client.close()
+
+
+def fetch_and_set_cors_origins_cache():
+    allow_origins = fetch_application_settings_cors_origins(
+        token=MOONSTREAM_ADMIN_ACCESS_TOKEN
+    )
+    set_cors_origins_cache(allow_origins)
+
+    return list(allow_origins)
+
+
+class BugoutCORSMiddleware(CORSMiddleware):
+    """
+    Modified CORSMiddleware from starlette.middleware.cors.py to work with Redis cache
+    and store application configuration for each user in Brood resources.
+    """
+
+    def __init__(
+        self,
+        app: ASGIApp,
+        allow_methods: Sequence[str] = ("GET",),
+        allow_headers: Sequence[str] = (),
+        allow_credentials: bool = False,
+        expose_headers: Sequence[str] = (),
+        max_age: int = 600,
+    ):
+        self.allow_origins = fetch_and_set_cors_origins_cache()
+
+        super().__init__(
+            app=app,
+            allow_origins=self.allow_origins,
+            allow_methods=allow_methods,
+            allow_headers=allow_headers,
+            allow_credentials=allow_credentials,
+            allow_origin_regex=None,
+            expose_headers=expose_headers,
+            max_age=max_age,
+        )
+
+    async def __call__(self, scope: Scope, receive: Receive, send: Send):
```

Comment:
I suggest:
1. If Redis connection is available, just check `SISMEMBER` on each request
2. If Redis connection is unavailable, check membership in application settings allowed origins on each request.

Why do we need to keep an `allow_origins` member around like this?

Reply:
We can read the hostname header on the request directly here and resolve CORS responses the same way that the original `CORSMiddleware` does. We don't need to maintain a separate `allow_origins` list if we can do a set membership check directly in Redis.

Reply:
We need to update `self.allow_origins` or totally rewrite this middleware class logic, what I am trying to avoid.